src/restconfig/src/main/java/org/geoserver/rest/resources/ResourceResource.java

```diff
@@ -0,0 +1,409 @@
+        xmlFormat.getXStream().processAnnotations(ResourceMetadata.class);
+        xmlFormat.getXStream().processAnnotations(ResourceDirectory.class);
+        xmlFormat.getXStream().aliasField("atom:link", ResourceReference.class, "link");
+        xmlFormat.getXStream().aliasField("atom:link", ResourceChild.class, "link");
+        formats.add(xmlFormat);
+        
+        //JSON
+        ReflectiveJSONFormat jsonFormat = new ReflectiveJSONFormat();
+        AtomLink.configureJSON(jsonFormat.getXStream());
+        jsonFormat.getXStream().processAnnotations(ResourceChild.class);
+        jsonFormat.getXStream().processAnnotations(ResourceMetadata.class);
+        jsonFormat.getXStream().processAnnotations(ResourceDirectory.class);
+        formats.add(jsonFormat);
+        
+        return formats;
+    }  
+
+    @Override
+    public void handleGet() {
+        if (operation == Operation.METADATA) {
+            getVariants().add(getFormatGet(false).toRepresentation(new ResourceMetadata(resource, this)));
+        }
+        else {
+            if (resource.getType() == Type.UNDEFINED) {
+                throw new RestletException("Undefined resource path.", Status.CLIENT_ERROR_NOT_FOUND);
+            } else {
+                Representation rep;
+                if (getRequest().getMethod().equals(Method.HEAD)) {
+                    rep = RESTUtils.emptyBody();
+                    rep.setMediaType(getMediaType(resource));
+                } else if (resource.getType() == Type.DIRECTORY) {
+                    rep = getFormatGet(false).toRepresentation(new ResourceDirectory(resource, this));
+                } else {
+                    rep = new InputRepresentation(resource.in(), getMediaType(resource));
+                }
+                rep.setModificationDate(new Date(resource.lastmodified()));
+                getVariants().add(rep);
+
+                if (!"".equals(resource.path())) {
+                    getResponseHeaders().add("Resource-parent", "/" + resource.parent().path());
+                }
+                getResponseHeaders().add("Resource-type", resource.getType().toString().toLowerCase());
+            }
+        }
+
+        super.handleGet();
+    }
+    
+    @Override
+    public boolean allowPut() {        
+        return true;        
+    }
+    
+    @Override
+    public void handlePut() {
+        if (resource.getType() == Type.DIRECTORY) {
+            throw new RestletException("Attempting to write data to a directory.", Status.CLIENT_ERROR_METHOD_NOT_ALLOWED);
+        } else {
+            try {
+                boolean isNew = resource.getType() == Type.UNDEFINED;
+                
+                if (operation == Operation.COPY || operation == Operation.MOVE) {                    
+                    try (InputStream is = getRequest().getEntity().getStream()) {
+                        Resource source = store.get(IOUtils.toString(is, "UTF-8"));
```

Comment:
@NielsCharlier I do not understand this. Why UTF-8? What happens if I PUT a resource with ```<?xml version="1.0" encoding="ISO-8859-15"?>``` in which paths contain non-ASCII characters?